File: majorProject/src/main/java/com/aocc/majorproject/EnemyController.java
package com.aocc.majorproject;

import java.util.LinkedList;
import java.util.Random;

import android.graphics.Paint;
import com.aocc.framework.Graphics;
import com.aocc.majorproject.GamePreferences;

// This controller allows one call in the GameScreen class to update or paint every 
// enemy through a LinkedList; it also allows the easy creation and deletion of enemies

public class EnemyController {
	private static final int MIN_SPAWN_TIME = 5;
	
	LinkedList<Enemy> e = new LinkedList<Enemy>();
	Enemy TempEnemy;
	Random r = new Random();
	
	private int nextEnemySpawn;
	private final GameSession session;

	public EnemyController(GameSession session) {
		this.session = session;
	}

	public int getNextEnemySpawn() {
		return nextEnemySpawn;
	}
	
	public void generateNextEnemy(int speed){
		nextEnemySpawn = r.nextInt(40 - 2*speed) + MIN_SPAWN_TIME;
	}
	
	// for adding and removing enemies from the LinkedList
	public void addEnemy(int x, int y, int t){
		TempEnemy = new Enemy(x, y, t, session);
		e.add(TempEnemy);
		this.generateNextEnemy(session.getSpeed());
	}
	
	public void removeEnemy(int i){
		e.remove(i);
		Assets.zap.play(GamePreferences.getTapVolume());
	}
	
	public void removeAllEnemies(){
		e.clear();
	}
	
	public void update(float deltaSeconds){
		for (int i = 0; i < e.size(); i++){
			TempEnemy = e.get(i);
			TempEnemy.update(deltaSeconds);
			if (TempEnemy.getHealth() <= 0) {
				this.removeEnemy(i);
			}
		}
	}

	public void increaseEnemyTopSpeed() {
		for (int i = 0; i < e.size(); i++) {
			e.get(i).increaseTopSpeed();
		}
	}
	
	public void paint(Graphics g, Paint paint){
		for (int i = 0; i < e.size(); i++){
			TempEnemy = e.get(i);
			TempEnemy.paint(g, paint);
		}
	}
	
	public LinkedList<Enemy> getE() {
		return e;
	}

	public void setE(LinkedList<Enemy> e) {
		this.e = e;
	}
	
}

File: majorProject/src/main/kotlin/com/aocc/majorproject/EnemyController.kt
package com.aocc.majorproject

import android.graphics.Paint
import android.graphics.PointF
import com.aocc.framework.Graphics
import java.util.LinkedList
import java.util.Random

/**
 * This controller allows one call in the GameScreen class to update or paint every
 * enemy through a LinkedList; it also allows the easy creation and deletion of enemies
 */
class EnemyController(private val session: GameSession) {

    var e: LinkedList<Enemy> = LinkedList()
    private val activeFormations = LinkedList<ActiveFormation>()
    private val r = Random()

    private var nextEnemySpawn = 0

    fun getNextEnemySpawn(): Int = nextEnemySpawn

    fun generateNextEnemy(speed: Int) {
        nextEnemySpawn = r.nextInt(40 - 2 * speed) + MIN_SPAWN_TIME
    }

    /** Default homing dot that tracks the player. */
    fun addEnemy(x: Int, y: Int, t: Int) {
        e.add(Enemy(x.toFloat(), y.toFloat(), t, session))
        generateNextEnemy(session.getSpeed())
    }

    /** Homing dot added without advancing the single-spawn cadence (for formations). */
    fun addTracking(x: Float, y: Float, t: Int) {
        e.add(Enemy(x, y, t, session, Enemy.Movement.TRACK))
    }

    /** Straight-line dot with no tracking; drifts across the screen and despawns off-edge. */
    fun addDrift(x: Float, y: Float, t: Int, vx: Float, vy: Float) {
        e.add(Enemy(x, y, t, session, Enemy.Movement.DRIFT, driftVx = vx, driftVy = vy))
    }

    /** Dot that follows a prescribed shape / trajectory, then drifts off and despawns. */
    fun addPath(x: Float, y: Float, t: Int, waypoints: List<PointF>, speed: Float) {
        e.add(Enemy(x, y, t, session, Enemy.Movement.PATH, path = waypoints, pathSpeed = speed))
    }

    /** Held dot whose position is driven by its formation until launched. Returns it. */
    fun addHeld(x: Float, y: Float, t: Int): Enemy {
        val enemy = Enemy(x, y, t, session, Enemy.Movement.HELD)
        e.add(enemy)
        return enemy
    }

    fun addFormation(formation: ActiveFormation) {
        activeFormations.add(formation)
    }

    fun removeEnemy(i: Int) {
        e.removeAt(i)
        Assets.zap?.play(GamePreferences.getTapVolume().toFloat())
    }

    fun removeAllEnemies() {
        e.clear()
        activeFormations.clear()
    }

    fun update(deltaSeconds: Float) {
        val formationIterator = activeFormations.iterator()
        while (formationIterator.hasNext()) {
            val formation = formationIterator.next()
            formation.update(deltaSeconds)
            if (formation.isComplete()) {
                formationIterator.remove()
            }
        }

        var i = 0
        while (i < e.size) {
            val enemy = e[i]
            enemy.update(deltaSeconds)
            when {
                enemy.getHealth() <= 0 -> removeEnemy(i)
                enemy.isDespawned() -> e.removeAt(i)
                else -> i++
            }
        }
    }

    fun increaseEnemyTopSpeed() {
        for (enemy in e) {
            enemy.increaseTopSpeed()
        }
    }

    fun paint(g: Graphics, paint: Paint) {
        for (enemy in e) {
            enemy.paint(g, paint)
        }
    }

    companion object {
        private const val MIN_SPAWN_TIME = 5
    }
}
